File: SpeakArena/Models/SpeakingModule.swift
import Foundation

struct SpeakingModule: Identifiable, Codable, Hashable {
    let id: UUID
    var title: String
    var subtitle: String
    var type: ModuleType

    // Learning path placement
    var pathIndex: Int
    var branch: Int // 0 = main trunk, +/- for side branches

    // Progression
    var status: ModuleStatus
    var requiredXP: Int
    var xpReward: Int
    var heartCost: Int

    // Difficulty
    var recommendedSkill: SkillLevel
    var estimatedDurationSeconds: Int

    // Optional unlock dependency
    var prerequisiteModuleID: UUID?

    init(
        id: UUID = UUID(),
        title: String,
        subtitle: String,
        type: ModuleType,
        pathIndex: Int,
        branch: Int = 0,
        status: ModuleStatus = .locked,
        requiredXP: Int = 0,
        xpReward: Int,
        heartCost: Int = 1,
        recommendedSkill: SkillLevel = .rookie,
        estimatedDurationSeconds: Int,
        prerequisiteModuleID: UUID? = nil
    ) {
        self.id = id
        self.title = title
        self.subtitle = subtitle
        self.type = type
        self.pathIndex = pathIndex
        self.branch = branch
        self.status = status
        self.requiredXP = requiredXP
        self.xpReward = xpReward
        self.heartCost = heartCost
        self.recommendedSkill = recommendedSkill
        self.estimatedDurationSeconds = estimatedDurationSeconds
        self.prerequisiteModuleID = prerequisiteModuleID
    }
}

File: backend/src/models/SpeakingModule.js
import mongoose from 'mongoose';

const speakingModuleSchema = new mongoose.Schema(
  {
    title: { type: String, required: true, trim: true },
    subtitle: { type: String, required: true, trim: true },
    type: {
      type: String,
      enum: ['p2p_debate', 'ai_sparring', 'elevator_pitch', 'articulation_drill'],
      required: true,
    },
    pathIndex: { type: Number, required: true, index: true },
    branch: { type: Number, default: 0 },
    status: { type: String, enum: ['locked', 'available', 'completed'], default: 'locked' },
    requiredXP: { type: Number, default: 0 },
    xpReward: { type: Number, required: true },
    heartCost: { type: Number, default: 1 },
    recommendedSkill: {
      type: String,
      enum: ['rookie', 'challenger', 'contender', 'orator', 'master'],
      default: 'rookie',
    },
    estimatedDurationSeconds: { type: Number, required: true },
    prerequisiteModuleId: { type: mongoose.Schema.Types.ObjectId, ref: 'SpeakingModule', default: null },
  },
  { timestamps: true }
);

speakingModuleSchema.index({ title: 1, type: 1 }, { unique: true });

export const SpeakingModule = mongoose.model('SpeakingModule', speakingModuleSchema);
